engine/src/state_chain_observer/sc_observer/mod.rs

```diff
@@ -292,306 +297,215 @@ where
-                                    }
-                                    state_chain_runtime::RuntimeEvent::EthereumVault(
-                                        pallet_cf_vaults::Event::KeyHandoverRequest {
-                                           ..
-                                        },
-                                    ) => {
-                                        panic!("There should be no key handover requests made for Ethereum")
-                                    }
-                                    state_chain_runtime::RuntimeEvent::PolkadotVault(
-                                        pallet_cf_vaults::Event::KeyHandoverRequest {
-                                           ..
-                                        },
-                                    ) => {
-                                        panic!("There should be no key handover requests made for Polkadot")
-                                    }
+                                        state_chain_client.clone(),
+                                        ceremony_id,
+                                        signatories,
+                                        signing_info,
+                                    ).await;
+                                }
+                    		}
+                            CfeEvent::EthKeygenRequest(req) => {
+                                handle_keygen_request::<_, _, _, EthereumInstance>(
+                                    scope,
+                                    &eth_multisig_client,
+                                    state_chain_client.clone(),
+                                    req.ceremony_id,
+                                    req.epoch_index,
+                                    req.participants,
+                                ).await;
+                            }
+                            CfeEvent::BtcKeygenRequest(req) => {
+                                handle_keygen_request::<_, _, _, BitcoinInstance>(
+                                    scope,
+                                    &btc_multisig_client,
+                                    state_chain_client.clone(),
+                                    req.ceremony_id,
+                                    req.epoch_index,
+                                    req.participants,
+                                ).await;
+                            }
+                            CfeEvent::DotKeygenRequest(req) => {
+                                handle_keygen_request::<_, _, _, PolkadotInstance>(
+                                    scope,
+                                    &dot_multisig_client,
+                                    state_chain_client.clone(),
+                                    req.ceremony_id,
+                                    req.epoch_index,
+                                    req.participants,
+                                ).await;
+                            }
```

Comment:
for clarity, I mean the duplication here (which I know existed before this PR) - would be nice if we can use the generics to dedup this somehow